Create draft for Unconference report (#53)
linebender/linebender.github.io: content/blog/2024-06-12-rustl-2024-unconference.md

+++
title = "Report on the RustNL 2024 Unconference"
authors = ["Olivier Faure"]
date = "2024-06-12 18:00:00"
+++

On May 7 and 8, 2024, folks from the Linebender team went to [the RustNL conference](https://2024.rustnl.org/) in Delft.
We listened to talks, Raph gave one, Rik Arends did the "hot reloading my entire editor live" thing that makes every other Rust GUI developer jealous, it was an all-around good time.

> (Videos for individual talks aren't up yet, but there are livestream recordings of [Day 1](https://www.youtube.com/watch?v=XLefuzE-ABU) and [Day 2](https://www.youtube.com/watch?v=521NfGf7AR0) on Youtube.)

Reviewer: I'd probably just remove the livestream links as well. I.e.
```suggestion
(Videos for individual talks are [on RustNL's YouTube channel](https://www.youtube.com/@rustnederlandrustnl/videos))
```
Or I'd even suggest just linking to the mentioned talks inline, and then removing this aside

After RustNL, though, another less-documented event called [The Unconference](https://2024.rustnl.org/unconf/). It was from roughly the same organizers and took place in the same city, but the format was fairly different.

> The Unconference wasn't livestreamed, and I haven't found any public discussion of it, so I thought I'd take the time to describe a bit of what was going.

Reviewer 2: ```suggestion
The Unconference wasn't livestreamed, and I haven't found any public discussion of it, so I thought I'd take the time to describe a bit of what was going on.
```

## The format

The "Unconference" was two days of loosely-structured chats between members of various Rust project teams where people talked about ecosystem collaboration, complained to Lang Team members that their favourite feature wasn't implemented yet, [went to pester everyone else about variadic generics](https://poignardazur.github.io/2024/05/25/report-on-rustnl-variadics/), and other good fun.

Attendees were split into three teams:

- The Rust project.
- GUI and Applications.
- Embedded development.

Discussions inside of these teams was mostly self-organized, with cross-team discussion mostly taking place during lunch and recesses.

### On the tyranny of struturelessness

*Note: This section is very much my personal opinion and doesn't represent the Linebender project as a whole.*

I'm not sure the self-organized format worked well.

From what I saw, the Embedded team took well to it, and the Rust team was kept productive thanks to Alice Cecile's efforts in marshalling everyone.

> In the case of the GUI team, people were spread in a very large room, which should have been conducive to small side-discussions and people splitting up to talk about the problems that interested them. Instead, there was an unspoken accord to progress through agenda items one by one, with a few people dominating the discussions on these items.

Reviewer 2: One sentence per line.

```suggestion
In the case of the GUI team, people were spread in a very large room, which should have been conducive to small side-discussions and people splitting up to talk about the problems that interested them.
Instead, there was an unspoken accord to progress through agenda items one by one, with a few people dominating the discussions on these items.
```

> Because those people we spread around a large room, they had to talk loudly to address each other, which left little room for side-discussions.

Reviewer 2: ```suggestion
Because those people were spread around a large room, they had to talk loudly to address each other, which left little room for side-discussions.
```

Our discussion were still productive, I just feel like the format could have been improved.

## The GUI team

As a member of the Linebender project, I was mostly exposed to the GUI team's proceedings.

The main topics of interest were:

- Build tooling (especially for Android and other non-desktop platforms).
- Text layout and editing.
- Winit adoption.
- Accessibility.
- Potential for further collaboration.
- Requests for new Rust features.

Note that I'm skipping a lot topics on the agenda, and other side discussions.
It feels like those topics were the most prominent during those two days.

There was also a fair amount of informal chatter around LogLog Games' strongly critical [Leaving Rust gamedev after 3 years](https://loglog.games/blog/leaving-rust-gamedev/) article.
I think those discussions mostly reflected the Bevy community's consensus, that the article had some flaws but overall pointed at very clear weaknesses in the Rust GUI / game engine ecosystem.

### Build tooling

Everyone agreed that a better solution was needed to ship non-TUI Rust apps.

> There was discussion about various platforms and build system, and the difficulty of having to eg write Java code for an Android port, and whether anyone wanted to take responsibility for maintaining that glue layer for the rest of the ecosystem.

Reviewer 2: ```suggestion
There was discussion about various platforms and build system, and the difficulty of having to e.g. write Java code for an Android port, and whether anyone wanted to take responsibility for maintaining that glue layer for the rest of the ecosystem.
```

No actionable decisions were made, but consensus seems to be that although integrated build tools are desirable, external build tooling for complex platforms (primarily Apple and Android) is likely to remain necessary.
Nobody seems to have stepped up to work on multi-target builds in Cargo, and since external build tools are required as it is, most people are okay with building everything several times when they release a multi-target package.
Most users are comfortable invoking platform-specific tools for bundling, signing, and manifest generation, so there is not much urgency for integrated tooling for these.

### Text layout and editing

There was some widespread interest in the Linebender ecosystem's text handling crates.

> From what I remember, Rik Arends and others had concerns about whether eg Parley and Swash depend on our Vello renderer (they don't).

Reviewer 2: ```suggestion
From what I remember, Rik Arends and others had concerns about whether e.g. Parley and Swash depend on our Vello renderer (they don't).
```

Some people also wanted Parley to be decoupled from Swash.

> People we especially interested in the idea of having access to a common text-editing widget, or at least infrastructure to create one.

Reviewer 2: ```suggestion
People were especially interested in the idea of having access to a common text-editing widget, or at least to infrastructure for creating one.
```

> That widget would handle IME, text selection, accessibility actions, etc, in a way that would feel native to different platforms with their own text-editing quirks (mobile, MacOS, Linux with vim mode, etc).

Reviewer 2: ```suggestion
That widget would handle IME, text selection, accessibility actions, etc, in a way that would feel native to different platforms with their own text-editing quirks (mobile, macOS, Linux with vim mode, etc).
```

### Winit

By now all of the Rust ecosystem has firmly converged on [`winit`](https://github.com/rust-windowing/winit) as the windowing platform-abstraction solution of choice.

(Well, not all! [One small project](https://github.com/makepad/makepad) with indomitable maintainers still holds out against the invaders.)

The consensus during this discussion seemed to be that Winit was there to stay.

> [TAO](https://github.com/tauri-apps/tao) users present were generally interested in being able to use upstream winit in the future, if its API adopts or supersedes the improvements made in TAO.

Reviewer 2: ```suggestion
[TAO](https://github.com/tauri-apps/tao) users present were generally interested in being able to use upstream `winit` in the future, if its API adopts or supersedes the improvements made in TAO.
```

Reviewer: Yeah, that or - at least, a scan of https://docs.rs/winit/latest/winit/ suggests it's now a proper noun; see "All of Winit’s functions"
```suggestion
[TAO](https://github.com/tauri-apps/tao) users present were generally interested in being able to use upstream Winit in the future, if its API adopts or supersedes the improvements made in TAO.
```

> Most were happy to see winit becoming more trait-oriented, and to see the `dpi` crate spun off from winit. 

Reviewer 2: ```suggestion
Most were happy to see `winit` becoming more trait-oriented, and to see the `dpi` crate spun off from winit. 
```

Glazier is still shelved and we're planning to port its features to Winit until it reaches feature parity.

A discussion that came up was the idea of splitting more "vocabulary types" from Winit into standalone crates.

> Vocabulary crates, eg crates that only define types and very little code, have a very low cost in memory and build times, while allowing projects to interop with eg Winit without actually adding a dependency on Winit.

Reviewer 2: ```suggestion
Vocabulary crates, e.g. crates that only define types and very little code, have a very low cost in memory and build times, while allowing projects to interop with e.g. Winit without actually adding a dependency on Winit.
```

This was connected with the discussions about text editing: having a vocabulary crate that would list Winit's events and also Parley's cursor movements would allow for a very low-dependency text editing widget.

### Accessibility

*Note: This section is also my personal opinion and doesn't represent the Linebender project either.*

The subject of Accessibility came up informally, but I'll claim it was underexplored.

While the subject was on the agenda, by the time we reached it the Unconference had been running long enough that everyone was tired and ready to leave or get to the catering.
This is an unfortunate pattern that I think is worth pointing out: if projects don't make a deliberate effort to prioritize discussions of accessibility, the subject will be left out.

Matt Campbell, the maintainer of AccessKit, could not attend the Unconference.
I'm concerned that, in his absence, other maintainers might have felt a dilution of responsibilty: since the accessibility guy wasn't there, nobody felt empowered to bring up accessibility.

> One point that did come up during the discussion is one I'd like to hammer in over the next few weeks, is that the Rust ecosystem should adopt a holistic view of accessibility.

Reviewer 2: ```suggestion
One point that did come up during the discussion which I'd like to hammer in over the next few weeks, is that the Rust ecosystem should adopt a holistic view of accessibility.
```

> Accessibility isn't just about screen readers, and making a framework accessible doesn't stop at adding a dependency to AccessKit, though doing so is a great first step.

Reviewer 2: ```suggestion
Accessibility isn't just about screen readers, and making a framework accessible doesn't stop at adding AccessKit as a dependency, though doing so is a great first step.
```

> Rik asked if there was a minimal set of the AccessKit elements that people could be expected to port their UI to.

Reviewer 2: If this was a separate discussion you had with Rik, then ignore this suggestion. But during the general unconference Rik wasn't talking about AccessKit, but about ARIA attributes.

```suggestion
Rik asked if there was a minimal set of ARIA attributes that people could start using in their UI, as the set of available attributes is gigantic.
```

The AccessKit framework is rich, and it's not obvious at first glance how much is needed versus nice-to-have.
The current reference in the Rust ecosystem is [the AccessKit implementation in eGUI](https://github.com/emilk/egui/pull/2294).

### Requests for Rust features

Jon Kelley, the creator of Dioxus, had [a laundry list of features](https://dioxus.notion.site/Dioxus-Labs-High-level-Rust-5fe1f1c9c8334815ad488410d948f05e) he wanted from the Rust language.
That list was later [filed in the Project Goals repository]((https://github.com/rust-lang/rust-project-goals/pull/10)).

The requests were well-documented, with motivating examples in existing Dioxus code that would clearly be improved by the features.

> Most of them felt both self-contained and like they would tremendously help beginners working on their first Rust project.

Reviewer 2: There were plenty of veterans at the unconference saying that the suggestions would help them personally, so I think it's misleading to paint these as beginner focused.

```suggestion
Most of them felt both self-contained and like they would tremendously help both beginners working on their first Rust project and long time veterans alike.
```

My favorite proposals were:

- `Capture` trait for auto-cloning reference-counted types in closures.
- Automatic partial borrows.
- Optional struct args.

Jon also wanted support for a remote cache of pre-built crates, so that users wouldn't suffer from having to rebuild dozens of crates every time you download your crate's dependencies.
The consensus from Rust maintainers was that such a cache would be essentially impossible with Rust's current architecture.

Niko Matsakis recently released [The borrow checker within](https://smallcultfollowing.com/babysteps/blog/2024/06/02/the-borrow-checker-within/), which feels like a follow-up to some of these discussions.

## Collaboration in the Rust GUI ecosystem

Over the last two years I've joked a few times that everybody in the Rust GUI ecosystem wants the ecosystem to standardize on common crates, as long as it's *their* crates.

After RustNL, I'm pleased to report this joke isn't accurate anymore.

A vibe that permeated all discussions was people wanting work to be pooled.
It felt like people didn't strongly care whose crate we adopted, as long as we all agreed the crate made sense:

> we're slowly crawling out of the xkcd 927 dilemma.

Reviewer 2: ```suggestion
we're slowly crawling out of the [xkcd 927](https://xkcd.com/927/) dilemma.
```
